### PR TITLE
Add TransformedMetricDataProvider.

### PR DESCRIPTION
**Pull request by mikelehen**
TransformedMetricDataProvider is a base class to help implement a MetricDataProvider that provides data by transforming data from another existing metric.  Implementers just need to set the `id` of the data provider and implement the `transformData()` method.

The idea is we can use this to create little data providers that compute daily new cases from cumulative cases, rolling averages, etc.  Example (note that it relies on a `Timeseries.rollingAverage()` method that's coming in a subsequent PR):

```
class RollingAverageMetricDataProvider extends TransformedMetricDataProvider {
  readonly id = "rolling-average";

  transformData(sourceData: MetricData<unknown>, newMetric: Metric, region: Region): MetricData<unknown> {
    const ts = sourceData.timeseries;
    const newTs = ts.rollingAverage({
      days: 7,
      treatMissingDatesAsZero: true
    });
    return new MetricData(newMetric, region, newTs.last?.value ?? null, newTs);
  }
}
```

Some additional incidental changes:
* Renamed MetricCatalog.fetchDataForMetricsAndRegions() to fetchDataForRegionsAndMetrics() for consistency.
* Made MetricCatalog pass itself to MetricDataProviders when fetching data so they can fetch
  dependent data if necessary.
* Added a MultiRegionMultiMetricDataStore.fromRegionsAndMetrics() helper for constructing a MultiRegionMultiMetricDataStore from regions,  metrics, and a MetricData getter callback.

Part of solving #89.

### CODE REVIEW

File: packages/metrics/src/MetricCatalog/MetricCatalog.test.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ describe("MetricCatalog", () => {
     const catalog = new MetricCatalog(testMetricDefs, dataProviders);
 
     // Fetch data for two metrics that come from different data providers.
-    const dataStore = await catalog.fetchDataForMetricsAndRegions(
+    const dataStore = await catalog.fetchDataForRegionsAndMetrics(
```
Comment (mikelehen) [inline packages/metrics/src/MetricCatalog/MetricCatalog.test.ts:91]: I renamed this for consistency with other methods (e.g. useDataForRegionsAndMetrics) and to match the order of the parameters you pass.

File: packages/metrics/src/MetricCatalog/MetricCatalog.ts [hunk 2]
```diff
@@ -186,7 +186,8 @@ export class MetricCatalog {
       const fetchedData = await provider.fetchData(
         regions,
         metrics,
-        includeTimeseries
+        includeTimeseries,
+        this
```
Comment (mikelehen) [inline packages/metrics/src/MetricCatalog/MetricCatalog.ts:190]: MetricCatalog now passes itself to `MetricDataProvider`s so they can fetch dependent metric data if necessary.

File: packages/metrics/src/data_providers/CachingMetricDataProviderBase.ts [hunk 3]
```diff
@@ -63,19 +59,11 @@ export abstract class CachingMetricDataProviderBase
...
-      data[region.regionId] = new MultiMetricDataStore(region, regionData);
-    }
-
-    return new MultiRegionMultiMetricDataStore(data);
+    return MultiRegionMultiMetricDataStore.fromRegionsAndMetrics(
+      regions,
+      metrics,
+      (region, metric) =>
+        this.getDataFromCache(region, metric, includeTimeseries)
+    );
```
Comment (mikelehen) [inline packages/metrics/src/data_providers/CachingMetricDataProviderBase.ts:67]: Refactored this to use the new helper.

File: packages/metrics/src/data_providers/MetricDataProvider.ts [hunk 4]
```diff
@@ -31,6 +32,7 @@ export interface MetricDataProvider {
   fetchData(
     regions: Region[],
     metrics: Metric[],
-    includeTimeseries: boolean
+    includeTimeseries: boolean,
+    metricCatalog: MetricCatalog
```
Comment (mikelehen) [inline packages/metrics/src/data_providers/MetricDataProvider.ts:40]: MetricDataProviders now receive the MetricCatalog.

File: packages/metrics/src/data/MultiRegionMultiMetricDataStore.ts [hunk 5]
```diff
@@ -50,6 +50,33 @@ export interface MultiRegionMultiMetricDataMap<T> {
 export class MultiRegionMultiMetricDataStore<T = unknown> {
   constructor(private data: MultiRegionMultiMetricDataMap<T>) {}
 
+  /**
```
Comment (mikelehen) [inline packages/metrics/src/data/MultiRegionMultiMetricDataStore.ts:53]: This code used to live in `CachingMetricDataProvider` but I needed to do the same sort of thing from `MetricTransformDataProvider` too, so I refactored it into a shared helper.

File: packages/metrics/src/data_providers/TransformedMetricDataProvider.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ * via the `sourceMetric` field of the `dataReference` field of the metric.
+ *
+ * @example
+ * ```typescript
+ * class RollingAverageMetricDataProvider extends TransformedMetricDataProvider {
+ *   readonly id = "rolling-average";
+ *
+ *   transformData(sourceData: MetricData<unknown>, newMetric: Metric, region: Region): MetricData<unknown> {
+ *     const ts = sourceData.timeseries;
+ *     const newTs = ts.rollingAverage({
```
Comment (mikelehen) [inline packages/metrics/src/data_providers/TransformedMetricDataProvider.ts:26]: rollingAverage() doesn't actually exist yet but is coming in another PR.

File: packages/metrics/src/data_providers/MetricDataProvider.ts (hunk 4 above)
Comment (smcclure17) [reply packages/metrics/src/data_providers/MetricDataProvider.ts:40]: super small nit: should we add a `@param`/description for `metricCatalog` here?

**Review by smcclure17 — APPROVED** [review]
This looks really great! Very impressed with how easy it will be to add new data providers off of this 🙏 🙌

File: packages/metrics/src/data_providers/MetricDataProvider.ts (hunk 4 above)
Comment (mikelehen) [reply packages/metrics/src/data_providers/MetricDataProvider.ts:40]: oops, yes!

**Pull request merged**